File: test/snippets/python/cyclewire.py
"""CycleWire attributes for server-rendered HTML: cw('cart#add', {'sku': sku}).

From https://github.com/CycleChain/CycleWire/blob/main/docs/server-helpers.md (MIT).
"""

import json
import re

_ACTION = re.compile(r'[A-Za-z0-9_.-]+(?:#[A-Za-z0-9_$]*)?')
_EVENT = re.compile(r'[a-z][a-z0-9:_-]*')
_PREFIX = re.compile(r'(?:[a-z0-9-]*-)?')
# Options printed after the action and props, in this order, with the strings
# they accept. debounce takes an integer instead, and once only True.
_OPTIONS = {
    'trigger': re.compile(r'load|idle|visible|media:.+', re.S),
    'preload': re.compile(r'intent|visible|idle|load|none'),
    'concurrency': re.compile(r'drop|restart|latest|parallel'),
    'debounce': None,
    'once': None,
    'prevent': re.compile(r'[a-z][a-z0-9:_-]*(?: [a-z][a-z0-9:_-]*)*'),
}
_ESCAPES = str.maketrans({'&': '&amp;', '"': '&quot;', "'": '&#39;', '<': '&lt;', '>': '&gt;'})


def _valid(name, value):
    if value is True:
        return name in ('once', 'prevent')
    if name == 'debounce':
        return type(value) is int and value >= 0  # not bool, which is an int too
    pattern = _OPTIONS[name]
    return isinstance(value, str) and pattern is not None and pattern.fullmatch(value) is not None


def cw(action, props=None, **options):
    """Return the attributes, escaped for HTML, to print inside a start tag.

    cw('cart#add', {'sku': 'wire-01'}, trigger='visible') returns
    cw-action="cart#add" cw-props="{&quot;sku&quot;:&quot;wire-01&quot;}" cw-trigger="visible"

    Raises ValueError for a bad name, option or value.
    """
    for name in options:
        if name not in _OPTIONS and name not in ('on', 'prefix'):
            raise ValueError(f'CycleWire: unknown option {name!r}')
    if not isinstance(action, str) or not _ACTION.fullmatch(action):
        raise ValueError(f'CycleWire: invalid action {action!r}')
    # None and False leave an option out, so values can come straight from variables.
    options = {k: v for k, v in options.items() if v is not None and v is not False}
    prefix = options.get('prefix', 'cw-')
    if not isinstance(prefix, str) or not _PREFIX.fullmatch(prefix):
        raise ValueError(f'CycleWire: invalid prefix {prefix!r}')
    on = options.get('on')
    if on is not None and (not isinstance(on, str) or not _EVENT.fullmatch(on)):
        raise ValueError(f'CycleWire: invalid on {on!r}')

    # An empty prefix means data-: a bare "action" is already a form attribute.
    base = prefix or 'data-'
    attributes = {base + ('action' if on is None else f'on-{on}'): action}
    if props is not None:
        # NaN and Infinity are not JSON, so they raise instead of breaking JSON.parse().
        attributes[f'{base}props'] = json.dumps(
            props, ensure_ascii=False, separators=(',', ':'), allow_nan=False
        )
    for name in _OPTIONS:
        if name not in options:
            continue
        value = options[name]
        if not _valid(name, value):
            raise ValueError(f'CycleWire: invalid {name} {value!r}')
        attributes[f'{base}{name}'] = None if value is True else str(value)
    return ' '.join(
        name if value is None else f'{name}="{value.translate(_ESCAPES)}"'
        for name, value in attributes.items()
    )

File: test/snippets/ruby/cycle_wire.rb
# frozen_string_literal: true

# CycleWire attributes for server-rendered HTML: CycleWire.cw('cart#add', { sku: sku }).
# From https://github.com/CycleChain/CycleWire/blob/main/docs/server-helpers.md (MIT).

require 'json'

module CycleWire
  ACTION = /\A[A-Za-z0-9_.-]+(?:#[A-Za-z0-9_$]*)?\z/
  EVENT = /\A[a-z][a-z0-9:_-]*\z/
  PREFIX = /\A(?:[a-z0-9-]*-)?\z/
  # Options printed after the action and props, in this order, with the strings
  # they accept. debounce takes an Integer instead, and once only true.
  OPTIONS = {
    trigger: /\A(?:load|idle|visible|media:.+)\z/m,
    preload: /\A(?:intent|visible|idle|load|none)\z/,
    concurrency: /\A(?:drop|restart|latest|parallel)\z/,
    debounce: nil,
    once: nil,
    prevent: /\A[a-z][a-z0-9:_-]*(?: [a-z][a-z0-9:_-]*)*\z/
  }.freeze
  ESCAPES = { '&' => '&amp;', '"' => '&quot;', "'" => '&#39;', '<' => '&lt;', '>' => '&gt;' }.freeze

  # The attributes, escaped for HTML, to print inside a start tag.
  # CycleWire.cw('cart#add', { sku: 'wire-01' }, trigger: 'visible') returns
  # data-cw-action="cart#add" data-cw-props="{&quot;sku&quot;:&quot;wire-01&quot;}" data-cw-trigger="visible"
  # Raises ArgumentError for a bad name, option or value.
  def self.cw(action, props = nil, options = {})
    options.each_key do |name|
      next if OPTIONS.key?(name) || name == :on || name == :prefix

      raise ArgumentError, "CycleWire: unknown option #{name.inspect}"
    end
    raise ArgumentError, "CycleWire: invalid action #{action.inspect}" unless action.is_a?(String) && ACTION.match?(action)

    # nil and false leave an option out, so values can come straight from variables.
    options = options.reject { |_, value| value.nil? || value == false }
    prefix = options.fetch(:prefix, 'cw-')
    raise ArgumentError, "CycleWire: invalid prefix #{prefix.inspect}" unless prefix.is_a?(String) && PREFIX.match?(prefix)

    on = options[:on]
    raise ArgumentError, "CycleWire: invalid on #{on.inspect}" unless on.nil? || (on.is_a?(String) && EVENT.match?(on))

    binds = on.nil? ? 'action' : "on-#{on}"
    attributes = { "data-#{prefix}#{binds}" => action }
    attributes["data-#{prefix}props"] = JSON.generate(props) unless props.nil?
    OPTIONS.each_key do |name|
      next unless options.key?(name)

      value = options[name]
      raise ArgumentError, "CycleWire: invalid #{name} #{value.inspect}" unless valid?(name, value)

      attributes["data-#{prefix}#{name}"] = value == true ? nil : value.to_s
    end
    attributes.map { |name, value| value.nil? ? name : %(#{name}="#{value.gsub(/[&"'<>]/, ESCAPES)}") }.join(' ')
  end

  def self.valid?(name, value)
    return %i[once prevent].include?(name) if value == true
    return value.is_a?(Integer) && value >= 0 if name == :debounce

    value.is_a?(String) && !OPTIONS[name].nil? && OPTIONS[name].match?(value)
  end
  private_class_method :valid?
end
